common/src/main/java/dev/latvian/mods/kubejs/item/custom/MultitoolItemJS.java

```diff
@@ -0,0 +1,198 @@
+package dev.latvian.mods.kubejs.item.custom;
+
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Multimap;
+import dev.latvian.mods.kubejs.item.MutableToolTier;
+import dev.latvian.mods.kubejs.registry.KubeJSRegistries;
+import dev.latvian.mods.kubejs.typings.Info;
+import dev.latvian.mods.kubejs.typings.Param;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.tags.BlockTags;
+import net.minecraft.tags.TagKey;
+import net.minecraft.world.InteractionResult;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.EquipmentSlot;
+import net.minecraft.world.entity.ai.attributes.Attribute;
+import net.minecraft.world.entity.ai.attributes.AttributeModifier;
+import net.minecraft.world.item.DiggerItem;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.Items;
+import net.minecraft.world.item.context.UseOnContext;
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.state.BlockState;
+import org.jetbrains.annotations.Nullable;
+
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+import static java.math.RoundingMode.CEILING;
+
+public class MultitoolItemJS extends DiggerItem {
+    public static class Builder extends HandheldItemBuilder {
+        protected final List<Float> speedValues = new ArrayList<>(5);
+        protected final List<Float> attackValues = new ArrayList<>(5);
+        protected final EnumSet<ToolTypes> toolTypes = EnumSet.noneOf(ToolTypes.class);
+
+        public Builder(ResourceLocation i) {
+            super(i, 0, 0);
+        }
+
+        @Info(value = """
+            Adds a tool to the multi-tool.
+            
+            Valid tool types include: 'axe', 'hoe', 'pickaxe', and 'shovel'.
+            """, params =
+        @Param(name = "tool", value = "The name of the tool to add to the multi-tool. Will error if it is an unknown tool type."))
+        public Builder tool(ToolTypes tool) {
+            toolTypes.add(tool);
+            attackValues.add(tool.attack);
+            speedValues.add(tool.speed);
+            return this;
+        }
+
+        @Override
+        public Builder attackDamageBaseline(float f) {
+            attackValues.clear();
+            attackValues.add(f);
+            return this;
+        }
+
+        @Override
+        public Builder speedBaseline(float f) {
+            speedValues.clear();
+            speedValues.add(f);
+            return this;
+        }
+
+        protected void setValues() {
+            // TOD0: find a way to make this better / more efficient
+            // i hate floating-point jank
+
```

Comment:
You could multiply by 10, round, cast to int, do the math with ints, then divide by 10f back to float. You will lose a tiny bit of precision, but you are already rounding later on.
It would still be messy though, not sure theres much you could do about that.

Reply:
im having a bit of trouble with this